File: built-in plugins dsp/mb-bass-reese.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbBassReesePlugin: PluginDefinition = {
    id: 'mb-bass-reese', slug: 'mb-bass-reese', name: 'MB Reese Bass', category: 'instrument', type: 'bass', version: '1.0.0',
    description: 'Detuned reese bass', author: 'Max Booster',
    oscillators: [{ type: 'sawtooth', detune: -7, gain: 0.5 }, { type: 'sawtooth', detune: 7, gain: 0.5 }],
    envelope: { attack: 0.01, decay: 0.3, sustain: 0.7, release: 0.3 },
    parameters: [
      { id: 'detune', name: 'Detune', type: 'float', defaultValue: 7, minValue: 0, maxValue: 30, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { detune: 7, volume: 0.8 },
  };

export default MbBassReesePlugin;

File: built-in plugins dsp/mb-bass-reese.cpp
/**
 * MB Reese Bass
 * Category : instrument
 * Type     : bass
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Detuned reese bass
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BASS_REESE_H
#define MB_BASS_REESE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBassReese : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-bass-reese";
    static constexpr const char* PLUGIN_NAME    = "MB Reese Bass";
    static constexpr const char* PLUGIN_TYPE    = "bass";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float detune = 7f;  // range [0, 30]
    float volume = 0.8f;  // range [0, 1]
    };

    MbBassReese() = default;
    ~MbBassReese() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.detune = std::clamp(params.detune, 0f, 30f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Reese Bass
        return input;
    }
};

#endif // MB_BASS_REESE_H
